File: ferrule/scheduler.py
"""Question selection: what should you practice next, and why.

Two ideas drive every choice.

1. Weakness-weighted sampling. Each skill gets a score combining how badly you
   perform on it, how much of the real exam it represents, and how little we
   know about you yet. Early sessions therefore behave like a diagnostic; later
   ones concentrate on whatever is actually costing you points.

2. Spaced repetition on misses. A question you got wrong comes back, first
   within the same sitting, then at widening gaps. Intervals are deliberately
   compressed relative to textbook SM-2 because people use this with a test
   weeks away, not months.
"""

import json
import math
import random

from . import db, sources


def _source_filter(conn, alias="q"):
    """Restrict a query to the sources the user practises from, and to the
    question vintage they asked for.

    Vintage matters because the bank keeps growing, and it grows in discrete
    batches rather than continuously -- College Board drops a few hundred
    questions at a time. The newest batch is the closest thing available to
    what the current exam looks like, which is worth practising on when your
    test is weeks away.

    Undated questions (the community source carries no dates) are excluded
    whenever a cutoff is active: "only questions added since July" cannot
    honestly include a question whose date nobody knows.
    """
    ids = sources.enabled_ids(conn)
    placeholders = ",".join("?" * len(ids))
    # `unusable` questions reference a figure the bank never shipped; they stay
    # browsable but must never be served as practice.
    clause = f"AND {alias}.source IN ({placeholders}) AND {alias}.unusable = 0"
    params = list(ids)

    cutoff = min_created(conn)
    if cutoff:
        clause += f" AND {alias}.created_at IS NOT NULL AND {alias}.created_at >= ?"
        params.append(cutoff)
    return clause, params


def min_created(conn):
    """Epoch-ms cutoff: only serve questions added at or after this instant."""
    raw = db.get_meta(conn, "min_created")
    try:
        return int(raw) if raw else 0
    except ValueError:
        return 0


def set_min_created(conn, cutoff_ms):
    db.set_meta(conn, "min_created", int(cutoff_ms or 0))
    return min_created(conn)

# Share of the real digital SAT each domain occupies. Used so that a weak but
# rare skill cannot crowd out a slightly-less-weak skill worth triple the points.
BLUEPRINT = {
    "Information and Ideas": 0.26,
    "Craft and Structure": 0.28,
    "Expression of Ideas": 0.20,
    "Standard English Conventions": 0.26,
    "Algebra": 0.35,
    "Advanced Math": 0.35,
    "Problem-Solving and Data Analysis": 0.15,
    "Geometry and Trigonometry": 0.15,
}

# Beta prior on per-skill accuracy. Centred slightly below average so an
# untested skill looks mildly risky and gets sampled early.
PRIOR_CORRECT = 1.6
PRIOR_TOTAL = 2.8

# Review intervals in days: same sitting, an hour, then spreading out.
INTERVALS = [0.01, 0.05, 0.5, 1.5, 3.0, 6.0]

MINUTE_MS = 60_000


def skill_stats(conn, test=None, source=None):
    """Per-skill accuracy, volume and pacing, blended with the prior."""
    where, params = "WHERE 1=1", []
    if test:
        where += " AND q.test = ?"
        params.append(test)
    if source:
        where += " AND q.source = ?"
        params.append(source)
    else:
        clause, sparams = _source_filter(conn)
        where += " " + clause
        params.extend(sparams)
    rows = conn.execute(
        f"""
        SELECT q.test, q.test_name, q.domain, q.skill,
               COUNT(a.id)                                   AS attempts,
               COALESCE(SUM(a.correct), 0)                   AS correct,
               COALESCE(AVG(NULLIF(a.elapsed_ms, 0)), 0)     AS avg_ms,
               MAX(a.answered_at)                            AS last_at
        FROM questions q
        LEFT JOIN attempts a ON a.external_id = q.external_id
        {where}
        GROUP BY q.test, q.domain, q.skill
        ORDER BY q.test, q.domain, q.skill
        """,
        params,
    ).fetchall()

    out = []
    for r in rows:
        attempts, correct = r["attempts"], r["correct"]
        est = (correct + PRIOR_CORRECT) / (attempts + PRIOR_TOTAL)
        out.append(
            {
                "test": r["test"],
                "test_name": r["test_name"],
                "domain": r["domain"],
                "skill": r["skill"],
                "attempts": attempts,
                "correct": correct,
                "accuracy": (correct / attempts) if attempts else None,
                "estimated": est,
                "avg_ms": int(r["avg_ms"] or 0),
                "last_at": r["last_at"],
            }
        )
    return out


def _priority(stat):
    """How much this skill deserves the next question. Higher wins."""
    weakness = (1.0 - stat["estimated"]) ** 1.6
    share = BLUEPRINT.get(stat["domain"], 0.2)
    # Uncertainty bonus decays as evidence accumulates -> explore, then exploit.
    uncertainty = 1.0 + 1.2 / math.sqrt(stat["attempts"] + 1.0)
    return max(weakness * share * uncertainty, 1e-6)


def target_difficulty(estimated):
    """Aim just above current ability: hard enough to teach, not to demoralise."""
    if estimated < 0.55:
        return ["E", "E", "M"]
    if estimated < 0.78:
        return ["M", "M", "E", "H"]
    return ["H", "H", "M"]


def due_reviews(conn, limit, test=None):
    where, params = "", []
    if test:
        where = "AND q.test = ?"
        params.append(test)
    clause, sparams = _source_filter(conn)
    where += " " + clause
    params.extend(sparams)
    params.append(db.now_ms())
    params.append(limit)
    return conn.execute(
        f"""
        SELECT q.* FROM reviews r
        JOIN questions q ON q.external_id = r.external_id
        WHERE q.stem IS NOT NULL AND q.stem != '' {where}
          AND r.due_at <= ?
        ORDER BY r.due_at ASC
        LIMIT ?
        """,
        params,
    ).fetchall()


def _unseen_for_skill(conn, skill, difficulties, exclude, allow_practice_test=False):
    """One unattempted question in this skill, preferring the target difficulty."""
    placeholders = ",".join("?" * len(exclude)) if exclude else "''"
    reserve = "" if allow_practice_test else "AND q.in_practice_test = 0"
    sclause, sparams = _source_filter(conn)
    for diff in difficulties + [None]:
        sql = f"""
            SELECT q.* FROM questions q
            LEFT JOIN attempts a ON a.external_id = q.external_id
            WHERE q.skill = ? AND q.stem IS NOT NULL AND q.stem != ''
              AND a.id IS NULL {reserve} {sclause}
              AND q.external_id NOT IN ({placeholders})
              {"AND q.difficulty = ?" if diff else ""}
            ORDER BY RANDOM() LIMIT 1
        """
        params = [skill] + sparams + list(exclude) + ([diff] if diff else [])
        row = conn.execute(sql, params).fetchone()
        if row:
            return row
    return None


def select_questions(
    conn, n=10, test=None, skill=None, review_share=0.35, allow_practice_test=False
):
    """Build the next practice set.

    By default this never serves a question that also appears in an official
    full-length practice test, so those stay unseen and your Bluebook scores
    remain a real measurement rather than a memory check.
    """
    picked, seen = [], set()
    reserve = "" if allow_practice_test else "AND q.in_practice_test = 0"
    sclause, sparams = _source_filter(conn)

    if skill:  # explicit drill: one skill, nothing else
        rows = conn.execute(
            f"""
            SELECT q.* FROM questions q
            LEFT JOIN attempts a ON a.external_id = q.external_id
            WHERE q.skill = ? AND q.stem IS NOT NULL AND q.stem != ''
              AND a.id IS NULL {reserve} {sclause}
            ORDER BY RANDOM() LIMIT ?
            """,
            [skill] + sparams + [n],
        ).fetchall()
        return [db.row_to_question(r) for r in rows]

    # Overdue misses first -- these are the highest-value questions in the bank.
    for row in due_reviews(conn, max(1, int(n * review_share)), test):
        picked.append(row)
        seen.add(row["external_id"])

    stats = [s for s in skill_stats(conn, test) if s["skill"]]
    if not stats:
        return [db.row_to_question(r) for r in picked]

    pool = list(stats)
    weights = [_priority(s) for s in pool]
    while len(picked) < n and pool:
        idx = random.choices(range(len(pool)), weights=weights, k=1)[0]
        stat = pool[idx]
        row = _unseen_for_skill(
            conn,
            stat["skill"],
            target_difficulty(stat["estimated"]),
            seen,
            allow_practice_test,
        )
        if row:
            picked.append(row)
            seen.add(row["external_id"])
        else:
            # This skill is exhausted (or not downloaded yet). Drop it, so a
            # partial question bank still yields a full set.
            pool.pop(idx)
            weights.pop(idx)

    if len(picked) < n:
        # Last resort: anything unseen at all, so the user is never handed a
        # short set just because their bank is incomplete.
        placeholders = ",".join("?" * len(seen)) if seen else "''"
        extra = conn.execute(
            f"""
            SELECT q.* FROM questions q
            LEFT JOIN attempts a ON a.external_id = q.external_id
            WHERE q.stem IS NOT NULL AND q.stem != '' AND a.id IS NULL
              {reserve} {sclause}
              AND q.external_id NOT IN ({placeholders})
              {"AND q.test = ?" if test else ""}
            ORDER BY RANDOM() LIMIT ?
            """,
            sparams + list(seen) + ([test] if test else []) + [n - len(picked)],
        ).fetchall()
        picked.extend(extra)

    random.shuffle(picked)
    return [db.row_to_question(r) for r in picked]


def grade(conn, external_id, response):
    """Compare a response against the stored key. Returns (correct, detail)."""
    row = conn.execute(
        "SELECT correct_answer, rationale, options, qtype FROM questions WHERE external_id = ?",
        (external_id,),
    ).fetchone()
    if not row:
        raise KeyError(external_id)

    keys = json.loads(row["correct_answer"] or "[]")
    given = (response or "").strip()

    if row["qtype"] == "spr":
        correct = any(_numeric_match(given, k) for k in keys)
    else:
        correct = given.upper() in {str(k).strip().upper() for k in keys}

    return correct, {
        "correct_answer": keys,
        "rationale": row["rationale"] or "",
        "qtype": row["qtype"],
    }


def _numeric_match(given, key):
    """Grid-in answers: accept equivalent numeric forms (0.5 == .5 == 1/2)."""
    a, b = str(given).strip(), str(key).strip()
    if not a:
        return False
    if a.lower() == b.lower():
        return True
    try:
        return math.isclose(_to_float(a), _to_float(b), rel_tol=1e-4, abs_tol=1e-6)
    except (ValueError, ZeroDivisionError):
        return False


def _to_float(s):
    s = s.replace(" ", "").replace(",", "")
    if "/" in s:
        num, _, den = s.partition("/")
        return float(num) / float(den)
    return float(s)


def record_attempt(conn, external_id, response, correct, elapsed_ms, session_id=None):
    """Log the attempt and advance this question's spaced-repetition state."""
    now = db.now_ms()
    conn.execute(
        """INSERT INTO attempts
           (external_id, session_id, answered_at, elapsed_ms, response, correct)
           VALUES (?, ?, ?, ?, ?, ?)""",
        (external_id, session_id, now, int(elapsed_ms or 0), response, 1 if correct else 0),
    )

    row = conn.execute(
        "SELECT * FROM reviews WHERE external_id = ?", (external_id,)
    ).fetchone()

    if not correct:
        # A miss always re-enters the queue, and repeat misses come back sooner.
        lapses = (row["lapses"] if row else 0) + 1
        ease = max(1.3, (row["ease"] if row else 2.4) - 0.22)
        interval = INTERVALS[0]
        reps = 0
    elif row:
        # A correct answer on a question you previously missed: step it along.
        reps = row["reps"] + 1
        ease = min(2.9, row["ease"] + 0.1)
        interval = INTERVALS[min(reps, len(INTERVALS) - 1)] * (ease / 2.4)
        lapses = row["lapses"]
    else:
        conn.commit()  # first-time correct: nothing to schedule
        return

    conn.execute(
        """
        INSERT INTO reviews (external_id, due_at, interval_days, ease, reps, lapses, last_at)
        VALUES (?, ?, ?, ?, ?, ?, ?)
        ON CONFLICT(external_id) DO UPDATE SET
            due_at = excluded.due_at, interval_days = excluded.interval_days,
            ease = excluded.ease, reps = excluded.reps,
            lapses = excluded.lapses, last_at = excluded.last_at
        """,
        (
            external_id,
            now + int(interval * 86_400_000),
            interval,
            ease,
            reps,
            lapses,
            now,
        ),
    )
    conn.commit()

File: web/lib/scheduler.js
/* Question selection: what should you practise next, and why.
 * Ported from ferrule/scheduler.py -- weakness-weighted sampling plus spaced
 * repetition on misses. The SQL is unchanged; see the Python for the reasoning
 * behind each weight and interval. */

import { getMeta, nowMs, rowToQuestion, setMeta } from './store.js';
import { enabledIds, placeholders } from './sources.js';

export function allowReserved(db) {
  return getMeta(db, 'allow_reserved') === '1';
}

export function setAllowReserved(db, on) {
  setMeta(db, 'allow_reserved', on ? '1' : '0');
  return allowReserved(db);
}

export function minCreated(db) {
  const raw = getMeta(db, 'min_created');
  const n = raw ? parseInt(raw, 10) : 0;
  return Number.isFinite(n) ? n : 0;
}

export function setMinCreated(db, cutoffMs) {
  setMeta(db, 'min_created', Math.trunc(Number(cutoffMs) || 0));
  return minCreated(db);
}

/* Restrict to enabled sources, usable questions, the reservation rule, and the
 * chosen vintage. Undated questions drop out whenever a cutoff is active. */
export function sourceFilter(db, alias = 'q') {
  const ids = enabledIds(db);
  let clause = `AND ${alias}.source IN (${placeholders(ids)}) AND ${alias}.unusable = 0`;
  const params = [...ids];
  if (!allowReserved(db)) clause += ` AND ${alias}.in_practice_test = 0`;
  const cutoff = minCreated(db);
  if (cutoff) {
    clause += ` AND ${alias}.created_at IS NOT NULL AND ${alias}.created_at >= ?`;
    params.push(cutoff);
  }
  return [clause, params];
}

// Share of the real digital SAT each domain occupies.
export const BLUEPRINT = {
  'Information and Ideas': 0.26,
  'Craft and Structure': 0.28,
  'Expression of Ideas': 0.20,
  'Standard English Conventions': 0.26,
  Algebra: 0.35,
  'Advanced Math': 0.35,
  'Problem-Solving and Data Analysis': 0.15,
  'Geometry and Trigonometry': 0.15,
};

const PRIOR_CORRECT = 1.6;
const PRIOR_TOTAL = 2.8;
const INTERVALS = [0.01, 0.05, 0.5, 1.5, 3.0, 6.0];

export function skillStats(db, test = null, source = null) {
  let where = 'WHERE 1=1';
  const params = [];
  if (test) { where += ' AND q.test = ?'; params.push(test); }
  if (source) {
    where += ' AND q.source = ?';
    params.push(source);
  } else {
    const [clause, sparams] = sourceFilter(db);
    where += ' ' + clause;
    params.push(...sparams);
  }
  return db.all(
    `SELECT q.test, q.test_name, q.domain, q.skill,
            COUNT(a.id)                                   AS attempts,
            COALESCE(SUM(a.correct), 0)                   AS correct,
            COALESCE(AVG(NULLIF(a.elapsed_ms, 0)), 0)     AS avg_ms,
            MAX(a.answered_at)                            AS last_at
     FROM questions q
     LEFT JOIN attempts a ON a.external_id = q.external_id
     ${where}
     GROUP BY q.test, q.domain, q.skill
     ORDER BY q.test, q.domain, q.skill`,
    params,
  ).map((r) => ({
    test: r.test,
    test_name: r.test_name,
    domain: r.domain,
    skill: r.skill,
    attempts: r.attempts,
    correct: r.correct,
    accuracy: r.attempts ? r.correct / r.attempts : null,
    estimated: (r.correct + PRIOR_CORRECT) / (r.attempts + PRIOR_TOTAL),
    avg_ms: Math.trunc(r.avg_ms || 0),
    last_at: r.last_at,
  }));
}

/* How much this skill deserves the next question. Higher wins. */
export function priority(stat) {
  const weakness = (1.0 - stat.estimated) ** 1.6;
  const share = BLUEPRINT[stat.domain] ?? 0.2;
  const uncertainty = 1.0 + 1.2 / Math.sqrt(stat.attempts + 1.0);
  return Math.max(weakness * share * uncertainty, 1e-6);
}

function targetDifficulty(estimated) {
  if (estimated < 0.55) return ['E', 'E', 'M'];
  if (estimated < 0.78) return ['M', 'M', 'E', 'H'];
  return ['H', 'H', 'M'];
}

function dueReviews(db, limit, test = null) {
  let where = '';
  const params = [];
  if (test) { where = 'AND q.test = ?'; params.push(test); }
  const [clause, sparams] = sourceFilter(db);
  where += ' ' + clause;
  params.push(...sparams, nowMs(), limit);
  return db.all(
    `SELECT q.* FROM reviews r
     JOIN questions q ON q.external_id = r.external_id
     WHERE q.stem IS NOT NULL AND q.stem != '' ${where}
       AND r.due_at <= ?
     ORDER BY r.due_at ASC
     LIMIT ?`,
    params,
  );
}

function unseenForSkill(db, skill, difficulties, exclude) {
  const ex = exclude.size ? placeholders([...exclude]) : "''";
  const [sclause, sparams] = sourceFilter(db);
  for (const diff of [...difficulties, null]) {
    const row = db.get(
      `SELECT q.* FROM questions q
       LEFT JOIN attempts a ON a.external_id = q.external_id
       WHERE q.skill = ? AND q.stem IS NOT NULL AND q.stem != ''
         AND a.id IS NULL ${sclause}
         AND q.external_id NOT IN (${ex})
         ${diff ? 'AND q.difficulty = ?' : ''}
       ORDER BY RANDOM() LIMIT 1`,
      [skill, ...sparams, ...exclude, ...(diff ? [diff] : [])],
    );
    if (row) return row;
  }
  return null;
}

function weightedIndex(weights) {
  const total = weights.reduce((a, b) => a + b, 0);
  let r = Math.random() * total;
  for (let i = 0; i < weights.length; i++) {
    r -= weights[i];
    if (r < 0) return i;
  }
  return weights.length - 1;
}

function shuffle(list) {
  for (let i = list.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [list[i], list[j]] = [list[j], list[i]];
  }
  return list;
}

/* Build the next practice set. */
export function selectQuestions(db, { n = 10, test = null, skill = null, reviewShare = 0.35 } = {}) {
  const picked = [];
  const seen = new Set();
  const [sclause, sparams] = sourceFilter(db);

  if (skill) { // explicit drill: one skill, nothing else
    return db.all(
      `SELECT q.* FROM questions q
       LEFT JOIN attempts a ON a.external_id = q.external_id
       WHERE q.skill = ? AND q.stem IS NOT NULL AND q.stem != ''
         AND a.id IS NULL ${sclause}
       ORDER BY RANDOM() LIMIT ?`,
      [skill, ...sparams, n],
    ).map(rowToQuestion);
  }

  // Overdue misses first -- the highest-value questions in the bank.
  for (const row of dueReviews(db, Math.max(1, Math.trunc(n * reviewShare)), test)) {
    picked.push(row);
    seen.add(row.external_id);
  }

  const stats = skillStats(db, test).filter((s) => s.skill);
  if (!stats.length) return picked.map(rowToQuestion);

  const pool = [...stats];
  const weights = pool.map(priority);
  while (picked.length < n && pool.length) {
    const idx = weightedIndex(weights);
    const stat = pool[idx];
    const row = unseenForSkill(db, stat.skill, targetDifficulty(stat.estimated), seen);
    if (row) {
      picked.push(row);
      seen.add(row.external_id);
    } else {
      // Skill exhausted (or not downloaded yet): drop it so a partial bank
      // still yields a full set.
      pool.splice(idx, 1);
      weights.splice(idx, 1);
    }
  }

  if (picked.length < n) {
    const ex = seen.size ? placeholders([...seen]) : "''";
    picked.push(...db.all(
      `SELECT q.* FROM questions q
       LEFT JOIN attempts a ON a.external_id = q.external_id
       WHERE q.stem IS NOT NULL AND q.stem != '' AND a.id IS NULL
         ${sclause}
         AND q.external_id NOT IN (${ex})
         ${test ? 'AND q.test = ?' : ''}
       ORDER BY RANDOM() LIMIT ?`,
      [...sparams, ...seen, ...(test ? [test] : []), n - picked.length],
    ));
  }
  return shuffle(picked).map(rowToQuestion);
}

function toFloat(s) {
  s = s.replace(/[ ,]/g, '');
  if (!/^[+-]?(\d+\.?\d*|\.\d+)(e[+-]?\d+)?(\/[+-]?(\d+\.?\d*|\.\d+))?$/i.test(s)) {
    throw new Error('not a number');
  }
  if (s.includes('/')) {
    const [num, den] = s.split('/');
    if (Number(den) === 0) throw new Error('division by zero');
    return Number(num) / Number(den);
  }
  return Number(s);
}

/* Grid-in answers: accept equivalent numeric forms (0.5 == .5 == 1/2). */
function numericMatch(given, key) {
  const a = String(given).trim();
  const b = String(key).trim();
  if (!a) return false;
  if (a.toLowerCase() === b.toLowerCase()) return true;
  try {
    const x = toFloat(a), y = toFloat(b);
    // math.isclose(rel_tol=1e-4, abs_tol=1e-6)
    return Math.abs(x - y) <= Math.max(1e-4 * Math.max(Math.abs(x), Math.abs(y)), 1e-6);
  } catch {
    return false;
  }
}

export function grade(db, externalId, response) {
  const row = db.get(
    'SELECT correct_answer, rationale, options, qtype FROM questions WHERE external_id = ?',
    [externalId],
  );
  if (!row) throw Object.assign(new Error(`unknown question '${externalId}'`), { status: 404 });
  const keys = JSON.parse(row.correct_answer || '[]');
  const given = (response || '').trim();
  const correct = row.qtype === 'spr'
    ? keys.some((k) => numericMatch(given, k))
    : keys.map((k) => String(k).trim().toUpperCase()).includes(given.toUpperCase());
  return [correct, { correct_answer: keys, rationale: row.rationale || '', qtype: row.qtype }];
}

/* Erase the most recent attempt at a question, for a misclick. */
export function undoAttempt(db, externalId) {
  const row = db.get(
    `SELECT id, correct FROM attempts WHERE external_id = ?
     ORDER BY answered_at DESC, id DESC LIMIT 1`,
    [externalId],
  );
  if (!row) return false;
  db.run('DELETE FROM attempts WHERE id = ?', [row.id]);
  const remaining = db.get('SELECT COUNT(*) AS n FROM attempts WHERE external_id = ?', [externalId]).n;
  if (remaining === 0) {
    db.run('DELETE FROM reviews WHERE external_id = ?', [externalId]);
  } else if (!row.correct) {
    db.run('UPDATE reviews SET lapses = MAX(0, lapses - 1) WHERE external_id = ?', [externalId]);
  }
  return true;
}

/* Log the attempt and advance this question's spaced-repetition state. */
export function recordAttempt(db, externalId, response, correct, elapsedMs, sessionId = null) {
  const now = nowMs();
  db.run(
    `INSERT INTO attempts
       (external_id, session_id, answered_at, elapsed_ms, response, correct)
     VALUES (?, ?, ?, ?, ?, ?)`,
    [externalId, sessionId ?? null, now, Math.trunc(Number(elapsedMs) || 0), response ?? null, correct ? 1 : 0],
  );
  const row = db.get('SELECT * FROM reviews WHERE external_id = ?', [externalId]);

  let lapses, ease, interval, reps;
  if (!correct) {
    // A miss always re-enters the queue, and repeat misses come back sooner.
    lapses = (row ? row.lapses : 0) + 1;
    ease = Math.max(1.3, (row ? row.ease : 2.4) - 0.22);
    interval = INTERVALS[0];
    reps = 0;
  } else if (row) {
    reps = row.reps + 1;
    ease = Math.min(2.9, row.ease + 0.1);
    interval = INTERVALS[Math.min(reps, INTERVALS.length - 1)] * (ease / 2.4);
    lapses = row.lapses;
  } else {
    return; // first-time correct: nothing to schedule
  }
  db.run(
    `INSERT INTO reviews (external_id, due_at, interval_days, ease, reps, lapses, last_at)
     VALUES (?, ?, ?, ?, ?, ?, ?)
     ON CONFLICT(external_id) DO UPDATE SET
         due_at = excluded.due_at, interval_days = excluded.interval_days,
         ease = excluded.ease, reps = excluded.reps,
         lapses = excluded.lapses, last_at = excluded.last_at`,
    [externalId, now + Math.trunc(interval * 86400000), interval, ease, reps, lapses, now],
  );
}
